### PR TITLE
Refactor ColSpec Interface

### PR DESCRIPTION
**Pull request by eliascarv**
This PR is a major refactoring of the ColSpec Interface. 
In this PR, instead of using Union Type to define ColSpec, it's using a Trait design.
`ColSpec` is now the parent type of all spec types. Are they:
* `NameSpec` - select columns using names
* `IndexSpec` - select columns using indices
* `RegexSpec` - select columns than match with regex
* `AllSpec` - select all columns
* `NoneSpec` - select no column

Beacuse of this new Trait design, a new step (step 2) has been added to implement the ColSpec Interface.

These are the new steps:

1 - Add colspec field:

```julia
struct MyTransform{S<:ColSpec,#= other type params =#}
  colspec::S
  # other fileds
end
```

2 - Convert spec to ColSpec using the `colspec` function:

```julia
MyTransform(spec, #= other arguments =#) = 
  MyTransform(colspec(spec), #= other arguments =#)
```

3 - Use `choose(colspec, names)` function in apply:

```julia
function apply(transform::MyTransform, table)
  cols = Tables.columns(table)
  names = Tables.columnnames(cols)
  # selected column names
  snames = choose(transform.colspec, names)
  # code...
end
```

If you need to create constructors that accept 
individual column selectors use the `Col` type. Example:

```julia
MyTransform(cols::T...) where {T<:Col} = 
  MyTransform(colspec(cols))
```

### CODE REVIEW

**Comment by eliascarv** [comment]
@juliohm, new files were added just to make the review simpler. They will be removed after review.

**Review by juliohm — CHANGES_REQUESTED** [review]
I think this PR is going in the right direction @eliascarv , really nice improvement over the previous design. Left a few review comments.

File: src/newcolspec.jl [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+ColSpec(1:10) # IndexSpec
+ColSpec((1, 2, 3)) # IndexSpec
+ColSpec(r"[abc]") # RegexSpec
+ColSpec(:) # AllSpec
+ColSpec(nothing) # NoneSpec
+# if the argument is a ColSpec, return it
+ColSpec(NoneSpec()) # NoneSpec
+```
+"""
+ColSpec(colspec::ColSpec) = colspec
```
[the remark this answers is not in the document]
Comment (eliascarv) [reply src/newcolspec.jl:102]: This style is inspired by the `Base.IndexStyle` and `Broadcat.BroadcatStyle` traits.

File: src/newcolspec.jl [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+
+choose(colspec::RegexSpec, names::Tuple) = choose(colspec, collect(names))
+function choose(colspec::RegexSpec, names::Vector)
+  regex = colspec.regex
+  snames = filter(nm -> occursin(regex, String(nm)), names)
+  @assert !isempty(snames) "Invalid column selection."
+  _choose(snames, names)
+end
+
+choose(::AllSpec, names::Tuple) = collect(names)
```
[the remark this answers is not in the document]
Comment (eliascarv) [reply src/newcolspec.jl]: @juliohm take a look at the new code order.

File: test/shows.jl [hunk 3]
```diff
@@ -4,27 +4,27 @@
 
     # compact mode
     iostr = sprint(show, T)
-    @test iostr == "Select((:a, :b, :c))"
+    @test iostr == "Select(TableTransforms.NameSpec([:a, :b, :c]))"
```
Comment (juliohm) [inline test/shows.jl:7]: We still need IO methods for these specs so that these tests remain unchanged.

**Review by juliohm — CHANGES_REQUESTED** [review]
Awesome @eliascarv , now we just need to add a Base.show method for the abstract ColSpec to show the field values instead.

File: test/shows.jl (hunk 3 above)
Comment (eliascarv) [reply test/shows.jl]: @juliohm takes a look at the new prints.

**Review by juliohm — APPROVED** [review]
That looks great now @eliascarv , feel free to replace the file and wait for the tests to pass.

**Pull request merged**